### PR TITLE
`ultralytics 8.3.200` Refactor Sony IMX exports to `imx.py`

### PR DESCRIPTION
**Pull request by Laughing-q**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Refactors and expands model export utilities: introduces Sony IMX500 export, renames ONNX/TensorRT helpers, streamlines IMX export flow, and updates docs — all bundled in version 8.3.200. 🚀

### 📊 Key Changes
- New IMX export module
  - Adds `ultralytics/utils/export/imx.py` with:
    - `torch2imx` for exporting models to Sony IMX500 format
    - `FXModel` (IMX-specific), `NMSWrapper`, and custom IMX decoding hooks
  - IMX export logic moved out of `Exporter` into reusable utilities
- API renames for clarity
  - `export_onnx` → `torch2onnx`
  - `export_engine` → `onnx2engine`
  - Updated imports/calls in `engine/exporter.py`
- Head module cleanup
  - Removes IMX-specific branches from `Detect`/`Pose` heads; IMX behavior now handled in the export module
- FXModel relocation
  - `FXModel` removed from `torch_utils.py` and moved/extended under IMX export (now accepts `imgsz`)
- Documentation updates
  - Replaces single export doc with a structured section:
    - [export/__init__.py reference](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/export/__init__.py)
    - [imx.py reference](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/export/imx.py)
  - Nav updated accordingly; removes `FXModel` from torch utils docs
- Version bump
  - `__version__` → 8.3.200

### 🎯 Purpose & Impact
- Clearer, modular export pipeline ✅
  - Easier maintenance and extension with dedicated IMX utilities
  - Cleaner model heads by moving IMX specifics to export-time
- Sony IMX500 deployment support 🎥
  - Quantization via MCT with built-in NMS and metadata export
  - Supports YOLOv8n and YOLO11n (detect and pose); includes label export
- More intuitive export APIs ⚙️
  - Function names now reflect direction: `torch2onnx`, `onnx2engine`, `torch2imx`
- Potential minor breaking change ⚠️
  - If you imported `export_onnx`/`export_engine` directly, update to `torch2onnx`/`onnx2engine`
- Better docs and discoverability 📚
  - Dedicated references for general export and Sony IMX paths

Quick usage examples:
- Install: `pip install ultralytics`
- High-level export:
  - ONNX: `yolo export format=onnx`
  - TensorRT: `yolo export format=engine`
  - IMX500: `yolo export format=imx` (Linux only; requires Sony IMX toolchain)
- Low-level helpers:
  - `from ultralytics.utils.export import torch2onnx, onnx2engine, torch2imx`

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Laughing-q, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response — an Ultralytics engineer will review and assist you shortly. To ensure a seamless integration of your work, please review the following checklist:

- ✅ Define a Purpose: Clearly explain the purpose of your fix or feature in your PR description, and link to any relevant issues(https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ Synchronize with Source: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ Ensure CI Checks Pass: Verify all Ultralytics Continuous Integration (CI)(https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ Update Documentation: Update the relevant documentation(https://docs.ultralytics.com/) for any new or modified features.
- ✅ Add Tests: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ Sign the CLA: Please ensure you have signed our Contributor License Agreement(https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ Minimize Changes: Limit your changes to the minimum necessary for your bug fix or feature addition. "It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is." — Bruce Lee

For more guidance, please refer to our Contributing Guide(https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

PR-specific notes for your IMX export refactor:
- Great move extracting IMX logic into `ultralytics/utils/export/imx.py` and exposing it via `ultralytics.utils.export` 🎯
- Thanks for cleaning the heads by removing IMX-specific branches and binding behavior via `FXModel` and `NMSWrapper` — this should simplify maintenance
- Please ensure:
  - All CI jobs pass, including style and docs builds ✅
  - New docs pages (`docs/en/reference/utils/export/__init__.md` and `docs/en/reference/utils/export/imx.md`) have complete descriptions/keywords instead of TODO placeholders
  - Backward compatibility for `exporter.export_imx()` behavior and metadata propagation is preserved
  - Tests cover both detection and pose flows for YOLO11/YOLOv8 with the new `export_imx` path, including labels file generation and metadata embedding
  - Linux and dependency checks remain consistent with prior behavior and error messages are informative
- The example usage you included is helpful 👍:
  ```python
  from ultralytics import YOLO

  model = YOLO("yolo11n.pt")
  model.export(format="imx", conf=0.25, iou=0.45, max_det=300)
  ```
- If there are known limitations (e.g., model variant constraints, required layer counts), consider surfacing them in the docs and PR summary for clarity 🧭

**Comment by Laughing-q** [comment]
@glenn-jocher @ambitious-octopus @onuralpszr as discussed, starting from imx export.
- Adding a standalone imx file to do all the dirty work, and eliminated some code in head.py we had to inject for imx, also include a standalone `export_imx` function.
- Creating a `export/` folder and put `export_onnx/engine` functions in the `__init__.py` for backward compatibility, not 100% sure about the structure though.
- @ambitious-octopus @lakshanthad I tested this branch with the intermediate onnx inference on both detect and pose tasks, both works correctly. Please double check on the imx device to make sure this branch works correctly there as well.

**Comment by glenn-jocher** [comment]
@Laughing-q this looks good but we have a naming conflict. I think let's update the Exporter class methods to start with underscores to avoid confusion, i.e. `Exporter._export_imx()`?

EDIT: Oh wait then I think they won't show up in the Docs if they are internal. Maybe just a normal name change then?

**Comment by Laughing-q** [comment]
@glenn-jocher yeah sounds good. Probably rename export_onnx/engine in `export/` to torch_to_onnx/onnx_to_engine, it makes this more readable as well as a standalone function, i.e. current `export_engine` function actually takes onnx file as input but no sign/info of that from the name of this function.
EDIT: ok updated, now the function names are `torch2onnx/onnx2engine/torch2imx`.

**Comment by glenn-jocher** [comment]
@Laughing-q yes this is better! Running slow CI



## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `break-exporter` at 2025-09-10 15:52:27 UTC with `@ultralytics/run-ci` command:

* ✅ [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/runs/17619395485): `ci.yml` run 11861
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/17619395707): `docker.yml` run 1114

**Comment by glenn-jocher** [comment]
@Laughing-q LGTM! @ambitious-octopus please review and merge if all good!

**Comment by lakshanthad** [comment]
@Laughing-q can confirm detection and post estimation works on IMX500!

**Comment by Laughing-q** [comment]
@lakshanthad awesome! thank you

**Review by ambitious-octopus — COMMENTED** [review]
@Laughing-q This looks good to me. Just thinking out loud here: what if we placed the export folder at the same level as the `engine` folder? That way, we'd have a completely dedicated module for exports, keeping the separation of concerns cleaner.

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `break-exporter` at 2025-09-15 15:20:26 UTC with `@ultralytics/run-ci` command:

* ✅ [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/runs/17738071720): `ci.yml` run 12040
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/17738071817): `docker.yml` run 1142

**Comment by UltralyticsAssistant** [comment]
🎉 Fantastic work team — excited to see this merged!

“Simplicity is the ultimate sophistication.” — Leonardo da Vinci. This PR embodies that spirit: a cleaner, modular export pipeline, intuitive APIs (`torch2onnx`, `onnx2engine`, `torch2imx`), and first-class Sony IMX500 support — all landing in 8.3.200. The streamlined heads and dedicated IMX utilities will make exports easier to maintain, extend, and ship across real-world deployments.

Huge thanks to @Laughing-q for leading, with valuable contributions from @lakshanthad, @glenn-jocher, and @ambitious-octopus — your craftsmanship shows.

Quick wins to try:
- CLI: `yolo export format=imx`
- Python:
  ```python
  from ultralytics.utils.export import torch2onnx, onnx2engine, torch2imx
  ```

Onward to smoother exports and broader edge deployments, including YOLOv8n and YOLO11n. Appreciate you all! 🚀
